engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,317 @@
+	if engine.IsUnverifiableInputError(err) || engine.IsOutdatedInputError(err) || engine.IsInvalidInputError(err) {
+		return nil
+	}
+
+	return err
+}
+
+// processApproval implements business logic for processing single approval
+// Returns:
+// * engine.InvalidInputError - result approval is invalid
+// * engine.UnverifiableInputError - result approval is unverifiable since referenced block cannot be found
+// * engine.OutdatedInputError - result approval is outdated for instance block was already sealed
+// * exception in case of any other error, usually this is not expected
+// * nil - successfully processed result approval
+func (c *approvalProcessingCore) processApproval(approval *flow.ResultApproval) error {
+	err := c.checkBlockOutdated(approval.Body.BlockID)
+	if err != nil {
+		return fmt.Errorf("won't process approval for oudated block (%x): %w", approval.Body.BlockID, err)
+	}
+
+	if collector, orphan := c.collectorTree.GetCollector(approval.Body.ExecutionResultID); collector != nil {
+		if orphan {
+			return engine.NewOutdatedInputErrorf("collector for %s is marked as orphan", approval.Body.ExecutionResultID)
+		}
+
+		// if there is a collector it means that we have received execution result and we are ready
+		// to process approvals
+		err = collector.ProcessApproval(approval)
+		if err != nil {
+			return fmt.Errorf("could not process assignment: %w", err)
+		}
+	} else {
+		// in case we haven't received execution result, cache it and process later.
+		c.approvalsCache.Put(approval)
+	}
+
+	return nil
+}
+
+func (c *approvalProcessingCore) checkEmergencySealing(lastSealedHeight, lastFinalizedHeight uint64) error {
+	if !c.emergencySealingActive {
+		return nil
+	}
+
+	emergencySealingHeight := lastSealedHeight + sealing.DefaultEmergencySealingThreshold
+
+	// we are interested in all collectors that match condition:
+	// lastSealedBlock + sealing.DefaultEmergencySealingThreshold < lastFinalizedHeight
+	// in other words we should check for emergency sealing only if threshold was reached
+	if emergencySealingHeight >= lastFinalizedHeight {
+		return nil
+	}
+
+	delta := lastFinalizedHeight - emergencySealingHeight
+	// if block is emergency sealable depends on it's incorporated block height
+	// collectors tree stores collector by executed block height
+	// we need to select multiple levels to find eligible collectors for emergency sealing
+	for _, collector := range c.collectorTree.GetCollectorsByInterval(lastSealedHeight, lastSealedHeight+delta) {
+		err := collector.CheckEmergencySealing(lastFinalizedHeight)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *approvalProcessingCore) processPendingApprovals(collector *AssignmentCollector) error {
+	// filter cached approvals for concrete execution result
+	for _, approvalID := range c.approvalsCache.ByResultID(collector.ResultID) {
+		if approval := c.approvalsCache.Take(approvalID); approval != nil {
```

Comment:
could we implement a `TakeByResultID`, which could eliminate the case where the `ByResultID` result will get a copy of approvals, and that copy would get stale if new approval arrives.

We probably don't need `ByResultID`. It seems the only chance we need to call `ByResultID` is when we want to take all approvals out for that result.